six/two/two.cpp

```diff
@@ -370,50 +370,71 @@ PyObject *Two::_findSubclassOf(PyObject *base, PyObject *module)
 
     PyObject *dir = PyObject_Dir(module);
     if (dir == NULL) {
+        PyErr_Clear();
         setError("there was an error calling dir() on module object");
         return NULL;
     }
 
-    PyObject *klass = NULL;
+    PyObject *klass;
     for (int i = 0; i < PyList_GET_SIZE(dir); i++) {
-        // get symbol name
-        char *symbol_name;
+        // Reset `klass` at every iteration so its state is always clean when we
+        // continue the loop or return early. Reset at first iteration is useless
+        // but it keeps the code readable.
+        klass = NULL;
+        Py_XDECREF(klass);
+
+        // get the symbol in current list item
         PyObject *symbol = PyList_GetItem(dir, i);
-        if (symbol != NULL) {
-            symbol_name = PyString_AsString(symbol);
+        if (symbol == NULL) {
+            // This should never happen as it means we're out of bounds
+            PyErr_Clear();
+            setError("there was an error browsing dir() output");
+            goto done;
+        }
+
+        // get symbol name
+        char *symbol_name = PyString_AsString(symbol);
+        if (symbol_name == NULL) {
+            // PyString_AsString returns NULL if `symbol` is not a string object
+            // and raises TypeError. Let's clear the error and keep going.
+            PyErr_Clear();
+            continue;
         }
 
         // get symbol instance. It's a new ref but in case of success we don't
-        // DecRef since we return it and the caller
-        // will be owner
+        // DecRef since we return it and the caller will be owner
         klass = PyObject_GetAttrString(module, symbol_name);
         if (klass == NULL) {
+            PyErr_Clear();
             continue;
         }
 
         // not a class, ignore
         if (!PyType_Check(klass)) {
```

Comment:
If the goal is to be extra-careful, shouldn't we also test that this call to `PyType_Check` returns -1 and if so do a `PyErr_Clear()` before continuing?
Same goes for `PyType_Check` line 360 & and `PyModule_Check` line 366.

Reply:
Those do not set an exception, see https://docs.python.org/2/c-api/type.html?highlight=pytype_check#c.PyType_Check and https://docs.python.org/2/c-api/module.html?highlight=pymodule_check#c.PyModule_Check.

In general, when you see `returns true if...` that means no error is set, while functions setting errors return `NULL`

Reply:
OK! I was following by the book the Tim Peters advice to also check for methods returning int.

Reply:
yeah, the thing about all the functions of the Py API is that the only way to know if they set errors/what's the return value when they set errors is to check their documentation :/ (even though in general the error value is `NULL`/`-1`)